Record timing stats

nicksellen opened the pull request:
From #1914 

My initial work for recording timing stats.

It works like this:
- we record various points using performance.mark/measure
- when we hit the first component that is mounted with the `v-measure` directive on it, we stop
- at that point we can collect up the points so far and do something with them (print them out in this case)
- we then clear again ready for the next route...

so there are two types of recordings:
1. initial page loading (this is the most interesting one)
2. subsequent page views within the SPA app (less interesting, but still interesting I think)

I called the point where we first render a `v-measure` component "MM" for "Meaningful Mount" (I didn't see a way to detect when it is actually painted, and a brief attempt to only trigger when it was visible showed that it was always visible by the time it was mounted, so I removed that bit). MM triggers earlier than Chromes FMP (First Meaningful Paint), looking at performance tools a bunch of stuff happens after our MM still, looks like re-rendering of stuff because dependent state changed (but not really... maybe just pointless re-renderings...). Still, I think the MM defines an interesting point we can try and reduce.

We could _just_ save MM, and not the intermediate stages, but maybe they are interesting too.

![performance](https://user-images.githubusercontent.com/31616/69907161-403baf00-13d0-11ea-9c33-2c2dd0886ebb.png)

TODO:
- [ ] make it work when `window.performance` is not available
- [ ] include whether it is initial request or not in payload
- [ ] send to backend
- [ ] tidy up a bit
- [ ] define `v-measure` in more places for other routes

nicksellen commented:
I decided to simplify it, so _only_ record the "meaningful mount" (the bit where the first `v-measure` is triggered). Will make it much simpler to graph and look at, and less data to store...

It's almost ready now... just need a backend, then to wire them up.

nicksellen commented:
So it records stuff like:

![simple](https://user-images.githubusercontent.com/31616/70047627-b276e400-15c8-11ea-9bd6-0459637fa997.png)

tiltec commented:
Nice! Maybe add the route params or the full path, that could make it easier to find out if some pages are slow because they have many variable elements on them.

nicksellen commented:
Yup, added that now too :)
So the stats backend will convert it into this format:

```js
[
  {
    "measurement": "karrot.stats.frontend",
    "fields": {
      "ms": 14661
    },
    "tags": {
      "first_load": true,
      "logged_in": true,
      "mobile": false,
      "group": 3,
      "route_name": "placePickups",
      "route_path": "/group/3/place/12/pickups",
      "route_params__group_id": "3",
      "route_params__place_id": "12"
    }
  }
]
```

nicksellen commented:
One additional thing I'm wondering whether to add in is the duration of various HTTP requests - as the current timing would not make it clear whether it's just a slow connection, or slow rendering.

... the performance API can tell me the duration of each of the resource requests (img, script, xhr, etc....), and maybe we could sum them per type. Many of the requests would happen in parallel but it would give an indication.

Or, we leave it till next time...

nicksellen commented:
Additionally, I think the beforeunload will not work in all browsers, but this sendBeacon thing is kind of designed for it and has more support: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon

nicksellen commented:
Hmm, send beacon doesn't allow to set custom headers (so server fails it with XSRF error), it could be disabled, but it looks like there is another approach using `keepalive` flag (https://fetch.spec.whatwg.org/#request-keepalive-flag), but can't see how to set that in axios. It has a `keepAlive` thing when using an agent, but that's a totally different thing iiuc.

Maybe this is enough as it is for now, and can do those bits later...

nicksellen commented:
I will make it work! https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests says:

> There are some cases in which the synchronous usage of XMLHttpRequest is not replaceable, like during the unload, beforeunload, and pagehide events. You should consider using the fetch() API with the keepalive flag. When fetch with keepalive isn't available, you can consider using the navigator.sendBeacon() API, which can support these use cases while typically delivering a good UX.

I've remade it with `fetch()` .. but will finish that up later...

nicksellen commented:
Ok, I added the resource timing too... just a total of everything. Should hopefully give some indication of connection speed.

tiltec commented:
I wonder if all browsers that support  the `performance` API also have `fetch` built in, then we wouldn't need the extra dependency!

nicksellen commented:
Could be! Also maybe I should load it conditionally depending on if fetch is available.

nicksellen commented:
Ok, I removed whatwg-fetch again, and we just only do the stats if we have a native fetch available. Should be enough browsers for that. fetch and performance browser support is not the same, some that support one but not the other.

I also added a few more stat fields whilst I was there, namely: `browser`, `os`, `dev`, `app`.

I think I'll stop now.

tiltec commented on src/utils/performance.js:
```diff
@@ -0,0 +1,145 @@
+  })
+}
+
+const debouncedSave = debounce(save, SAVE_INTERVAL_MS, { maxWait: SAVE_INTERVAL_MS * 2 })
```
Or rather use `debounceAndFlushBeforeUnload` and remove the `unload` listener at the end of the file?

nicksellen replied:
Ah yeah, I went in a few cycles with it, and back to the starting point again :)
